src/accountsdb/manager.zig

```diff
@@ -20,260 +20,201 @@ const AccountFile = sig.accounts_db.accounts_file.AccountFile;
         };
 
-        const slots_per_incremental_snapshot = self.config.slots_per_incremental_snapshot;
+        if (make_inc_snapshot and has_made_full_snapshot) {
+            std.debug.assert(!make_full_snapshot);
 
-        if (largest_flushed_slot - latest_inc_snapshot_slot >= slots_per_incremental_snapshot and
-            largest_flushed_slot - latest_full_snapshot_slot >= slots_per_incremental_snapshot)
-        {
-            self.db.logger.info().logf(
+            db.logger.info().logf(
                 "accountsdb[manager]: generating incremental snapshot from {d} to {d}",
-                .{ latest_full_snapshot_slot, largest_flushed_slot },
+                .{ latest_full_snapshot_slot, newly_rooted_slot },
             );
-            _ = try self.db.generateIncrementalSnapshotWithCompressor(
-                self.zstd_compressor,
-                self.zstd_buffer,
+            const result = try db.generateIncrementalSnapshotWithCompressor(
+                zstd_compressor,
+                zstd_buffer,
                 .{
-                    .target_slot = largest_flushed_slot,
-                    .bank_fields = &self.tmp_bank_fields,
-                    .lamports_per_signature = self.prng.random().int(u64),
+                    .target_slot = newly_rooted_slot,
+                    .bank_fields = &tmp_bank_fields,
+                    .lamports_per_signature = lamports_per_signature,
                     .old_snapshot_action = .delete_old,
                 },
             );
-        }
 
-        if (must_flush_slots) {
-            // clean the flushed slots account files
-            const clean_result = try cleanAccountFiles(
-                allocator,
-                self.db,
-                latest_full_snapshot_slot,
-                self.unclean_account_files.items,
-                &self.shrink_account_files,
-                &self.delete_account_files,
-            );
-            _ = clean_result;
-            // self.db.logger.debug().logf("clean_result: {any}", .{clean_result});
-
-            // shrink any account files which have been cleaned
-            const shrink_result = try shrinkAccountFiles(
-                allocator,
-                self.db,
-                self.shrink_account_files.keys(),
-                &self.delete_account_files,
-            );
-            _ = shrink_result;
-            // self.db.logger.debug().logf("shrink_results: {any}", .{shrink_results});
+            const maybe_gen_info: *?SnapshotGenerationInfo, var gen_info_lg =
+                db.latest_snapshot_gen_info.writeWithLock();
+            defer gen_info_lg.unlock();
 
-            // delete any empty account files
-            defer self.delete_account_files.clearRetainingCapacity();
-            try deleteAccountFiles(self.db, self.delete_account_files.keys());
+            const gen_info = &((maybe_gen_info.*) orelse
+                @panic("illegal state - snapshot_gen_info (previously non-null) is now null"));
```

Comment:
When could this happen? If this function is called twice concurrently?

Reply:
I'm not aware of any place this could happen - just programming defensively here (and preferring panic over unreachable).